### PR TITLE
Convert to typescript

### PR DESCRIPTION
**Pull request by antonnyman**
In this PR I want to convert the entire lib to Typescript. Compilation is handled with `microbundle`. The API should stay the same.

\+ Converted entire lib to class based Typescript.
\+ Added examples in the `/examples`-folder (based on stimulus).
\+ Now you need to register your controllers after creating the app, before we just called them from the window.
\+ Added README with examples.

\- Removed JS files
\- Removed DRS (Drag Reduction System 🏎️ , a pjax router exploration). Will create an issue.

🐞 Multiple targets with the same name is not supported yet. https://github.com/antonnyman/embroidery/issues/8

\? When building I get the error `(terser plugin) SyntaxError: Unexpected token: punc (.)`. I have no idea what this means. But it seems to build anyway.

### CODE REVIEW

File: lib/controller.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+type Context = {
+  [key: string]: Function
+}
+
+export default class Controller {
+  private dataAttr: string
+  private targetElements: NodeList
+  private actionElements: NodeList
+
+  private cache = {
```
Comment (antonnyman) [inline lib/controller.ts:12]: Cache might be a bad name here. Any suggestion on terms?

File: lib/index.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+type Callback = (el: Element) => void
+
+type Cache = {
+  [DataAttribute.Controller]: Element[]
+  [DataAttribute.Partial]: Element[]
+}
+
+export class Embroidery {
+  private cache: Cache = {
```
Comment (Skalman) [inline lib/index.ts:23]: Use readonly when possible

```suggestion
  private readonly cache: Cache = {
```

File: lib/index.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+  private context
+
+  static start() {
+    const app = new Embroidery()
+    app.start()
+    return app
+  }
+
+  constructor() {}
```
Comment (Skalman) [inline lib/index.ts:36]: Remove unused constructor
```suggestion
```

File: lib/index.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      .querySelectorAll(type)
+      .forEach((controller: Element) => callback(controller))
+  }
+
+  updateCache(element, type) {
+    this.cache = { ...this.cache, [type]: [...this.cache[type], element] }
+  }
+
+  initialize(e: Element) {
+    const element = e as HTMLElement
```
Comment (Skalman) [inline lib/index.ts:70]: Why not require an `HTMLElement` as the argument, if that's what you need?

File: lib/index.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+              this.updateCache(element, DataAttribute.Partial)
+              return new Partial(element)
+
+            default:
+              throw new Error(
+                'Element is not a specified data type, like controller or partial. Did you forget to register it?'
+              )
+          }
+        })
+      } catch (error) {
```
Comment (Skalman) [inline lib/index.ts:89]: Interesting. Any reason why you need to catch the error?

File: lib/index.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  discoverUninitializedControllers(
+    callback: Callback,
+    parentElement: Element = null
+  ) {
+    return Array.from(
+      (parentElement || document).querySelectorAll(Data.Controller)
+    )
+      .filter((element) => !this.cache[Data.Controller]?.includes(element))
```
Comment (Skalman) [inline lib/index.ts:108]: I think you can remove the `?`
```suggestion
      .filter((element) => !this.cache[Data.Controller].includes(element))
```

File: lib/index.ts [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+  listenForNewUninitializedControllersAtRuntime(callback: Callback) {
+    return new MutationObserver((mutations) =>
+      mutations.map(({ addedNodes }) => {
+        if (addedNodes.length > 0) {
+          addedNodes.forEach((node) => {
+            // Ignore non-elements
+            if (node.nodeType !== 1) return
+            // Don't listen for changes that happen in discovered controllers
+            if (node?.parentElement?.closest(Data.Controller)) return
```
Comment (Skalman) [inline lib/index.ts:120]: `node` cannot be null/undefined
```suggestion
            if (node.parentElement?.closest(Data.Controller)) return
```

File: lib/index.ts [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+      (parentElement || document).querySelectorAll(Data.Controller)
+    )
+      .filter((element) => !this.cache[Data.Controller]?.includes(element))
+      .map((element) => callback(element))
+  }
+
+  listenForNewUninitializedControllersAtRuntime(callback: Callback) {
+    return new MutationObserver((mutations) =>
+      mutations.map(({ addedNodes }) => {
+        if (addedNodes.length > 0) {
```
Comment (Skalman) [inline lib/index.ts:115]: You can remove this `if`

File: lib/index.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  ) {
+    return Array.from(
+      (parentElement || document).querySelectorAll(Data.Controller)
+    )
+      .filter((element) => !this.cache[Data.Controller]?.includes(element))
+      .map((element) => callback(element))
+  }
+
+  listenForNewUninitializedControllersAtRuntime(callback: Callback) {
+    return new MutationObserver((mutations) =>
```
Comment (Skalman) [inline lib/index.ts:113]: You aren't returning anything anyway, so you might as well remove the `return`
```suggestion
    new MutationObserver((mutations) =>
```

File: lib/partial.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  constructor(element: Element) {
+    this.element = element
+    this.dataAttr = element.getAttribute('data-partial')
+
+    this.create()
+  }
+
+  async create() {
+    const result = await fetch(this.dataAttr, {
+      method: 'GET',
```
Comment (Skalman) [inline lib/partial.ts:15]: You can remove this. `GET` is already the default
```suggestion
```

File: lib/controller.ts (hunk 1 above)
Comment (Skalman) [reply lib/controller.ts:12]: Can't think of a better one. Good enough, I think.

But you should make things readonly if possible  :)

```suggestion
  private readonly cache = {
```

Comment (antonnyman) [reply lib/controller.ts:12]: Ok, will go with cache for this one.
In this case cache is not readonly (see the updateCache method). :)

**Comment by antonnyman** [comment]
Some files were deleted so this PR had to be merged manually.

**Pull request closed without merging**